scripts/utils.py

```diff
@@ -16,28 +17,34 @@ def extract_tar_file_contents(filename, filetype):
     """
     extension = EXTENSION_BY_FILETYPE[filetype]
 
-    tar = tarfile.open(filename)
-    internal_file = None
-    for member in tar.getmembers():
-        suffixes = Path(member.name).suffixes
+    with tarfile.open(filename) as tar:
+        internal_file = None
+        for member in tar.getmembers():
+            suffixes = Path(member.name).suffixes
 
-        if extension in suffixes:
-            # By default, return the binary stream for the member file.
-            internal_file = tar.extractfile(member.name)
+            if extension in suffixes:
+                # By default, return the binary stream for the member file.
+                internal_file = tar.extractfile(member.name)
 
-            if ".xz" in suffixes:
-                # Check for LZMA-compressed data and open these with the
-                # corresponding library.
-                internal_file = lzma.open(internal_file, "rt")
-            elif extension == ".fasta":
-                # For sequence data, handle decoding of the binary stream prior
-                # to passing the data back to the caller.
-                internal_file = (line.decode("utf-8") for line in internal_file)
+                if ".xz" in suffixes:
+                    # Check for LZMA-compressed data and open these with the
+                    # corresponding library.
+                    internal_file = lzma.open(internal_file, "rt")
 
-            break
+                break
 
-    if internal_file is None:
-        tar.close()
-        raise FileNotFoundError(f"Could not find a {filetype} file in '{filename}'")
+        if internal_file is None:
+            raise FileNotFoundError(f"Could not find a {filetype} file in '{filename}'")
 
-    return internal_file, tar
+        # Write the internal file out to an uncompressed temporary file. This
+        # approach allows downstream processes to re-read the file in multiple
+        # passes instead of making a single pass through a stream.
+        with NamedTemporaryFile(delete=False) as temporary_file:
+            temporary_file_path = temporary_file.name
+            for line in internal_file:
+                if not isinstance(line, bytes):
+                    line = line.encode("utf-8")
+
+                temporary_file.write(line)
```

Comment:
I don't understand why this iterates over `internal_file` writing line-by-ilne (from what it expects might be binary handles, even) to `temporary_file`. I'd expect to see `shutil.copyfileobj()` or `tarfile.extract()` instead. Can you explain it to me? If the reason is only for the conditional UTF-8 encoding, then see my comment about lzma handling above.